### PR TITLE
Replace dynamic_cast<>

### PR DESCRIPTION
**Pull request by siying**
Summary:Replace dynamic_cast<> so that users can choose to build with RTTI off, so that they can save several bytes per object, and get tiny more memory available.
Some nontrivial changes:
1. Add Comparator::GetRootComparator() to get around the internal comparator hack
2. Add the two experiemental functions to DB
3. Add TableFactory::GetOptionString() to avoid unnecessary casting to get the option string
4. Since 3 is done, move the parsing option functions for table factory to table factory files too, to be symmetric.

Test Plan:Run all existing tests

### CODE REVIEW

File: cache/clock_cache.cc [hunk 1]
```diff
@@ -27,6 +27,9 @@ std::shared_ptr<Cache> NewClockCache(size_t capacity, int num_shard_bits,
 #include <atomic>
 #include <deque>
 
+// "tbb/concurrent_hash_map.h" requires RTTI if exception is enabled.
+// Disable it so users can chooose to disable RTTI.
+#define TBB_USE_EXCEPTIONS 0
```
Comment (IslamAbdelRahman) [inline cache/clock_cache.cc]: Do we want to put this inside
```
#ifndef USE_RTTI
```

File: examples/compaction_filter_example.cc [hunk 2]
```diff
@@ -59,7 +59,10 @@ int main() {
 
   MyFilter filter;
 
-  system("rm -rf /tmp/rocksmergetest");
+  int ret = system("rm -rf /tmp/rocksmergetest");
+  if (ret != 0) {
+    fprintf(stderr, "Error deleting /tmp/rocksmergetest, code: %d\n", ret);
```
Comment (IslamAbdelRahman) [inline examples/compaction_filter_example.cc]: nit: maybe we should also add
```
return ret;
```

File: utilities/transactions/optimistic_transaction_impl.cc [hunk 3]
```diff
@@ -122,8 +116,7 @@ Status OptimisticTransactionImpl::TryLock(ColumnFamilyHandle* column_family,
 Status OptimisticTransactionImpl::CheckTransactionForConflicts(DB* db) {
   Status result;
 
-  assert(dynamic_cast<DBImpl*>(db) != nullptr);
```
Comment (IslamAbdelRahman) [inline utilities/transactions/optimistic_transaction_impl.cc]: I guess we can still keep these asserts

File: tools/sst_dump_tool.cc [hunk 4]
```diff
@@ -14,6 +14,7 @@
 #include <inttypes.h>
 #include <iostream>
 #include <map>
+#include <memory>
```
Comment (IslamAbdelRahman) [inline tools/sst_dump_tool.cc]: Just wondering, why ?

**Comment by siying** [comment]
I thought about doing dynamic_cast in NDEBUG but I finally think it is good to have DEBUG to disable RTTI too, to make sure debug and opt are close in this sense.

File: tools/sst_dump_tool.cc (hunk 4 above)
Comment (siying) [inline tools/sst_dump_tool.cc]: This is very weird. If I don't add this, shared_ptr<> used will be a wrong one which include a typeinfo call. Don't get why but it's a good habit to add it anyway, as the example: http://en.cppreference.com/w/cpp/memory/shared_ptr

**Comment by siying** [comment]
@IslamAbdelRahman I've addressed the comments.

File: tools/db_bench_tool.cc [hunk 5]
```diff
@@ -2551,7 +2552,7 @@ void VerifyDBFromDB(std::string& truth_db_name) {
     }
     if (FLAGS_simcache_size >= 0) {
       fprintf(stdout, "SIMULATOR CACHE STATISTICS:\n%s\n",
-              std::dynamic_pointer_cast<SimCache>(cache_)->ToString().c_str());
+              static_cast<SimCache*>(cache_.get())->ToString().c_str());
```
Comment (IslamAbdelRahman) [inline tools/db_bench_tool.cc]: _with_check ?

**Pull request closed without merging**